### PR TITLE
pkg/ipnet: Add DeepCopy(Into) support to IPNet

### PR DESCRIPTION
**Pull request by wking**
@dgoodwin [needs `DeepCopyInto` on `IPNet`][1] as part of embedding our types into Hive's `ClusterDeployment` CRD.  The `zz_generated.deepcopy` prefix matches the current Kubernetes pattern, namespaces us from other future generators besides deepcopy, and sorts the file after human-generated entries in collated directory listings.

`deepcopy-gen` only looks for the package tag in file-level comments in a `doc.go` file (kubernetes/gengo#124).  I think that's an unfortunate restriction, but with this pull request I'm playing along by moving our package doc over to a new `doc.go`.  I'm also adding [the tag that deepcopy-gen is expecting][3].  And I'm also adding [a `go:generate` tag][4] that will run `deepcopy-gen` for us.

I expect this generated output to be pretty stable, which gives us some time to work up CI tooling for this approach.  Once we have CI coverage to protect from folks forgetting to regenerate, I'll extend this approach to cover more of the `InstallConfig` types.  But focusing on `IPNet` seemed like a reasonable starting point for testing the waters.

[1]: https://github.com/openshift/installer/issues/256#issue-360875062
[2]: https://github.com/kubernetes/gengo/pull/124
[3]: https://github.com/kubernetes/gengo/blob/4242d8e6c5dba56827bb7bcf14ad11cda38f3991/examples/deepcopy-gen/main.go#L33-L36
[4]: https://golang.org/pkg/cmd/go/internal/generate/

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
https://blog.openshift.com/kubernetes-deep-dive-code-generation-customresources/
Pins the generator to a specific version in vendor directory; that is important.

go generate doesn't allows to pin generator's version.

File: pkg/ipnet/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+limitations under the License.
+*/
+
+// Code generated by deepcopy-gen. DO NOT EDIT.
+
+package ipnet
+
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *IPNet) DeepCopyInto(out *IPNet) {
+	*out = *in
```
Comment (crawford) [inline pkg/ipnet/zz_generated.deepcopy.go]: This is a pretty incredible amount of complexity for this single line of code. Can we just use a simplified version of this implementation and drop all of the terrible k8s code generation?

Comment (wking) [reply pkg/ipnet/zz_generated.deepcopy.go]: > Can we just use a simplified version of this implementation and drop all of the terrible k8s code generation?

I'm fine going that way to.  Auto-generation is at @dgoodwin's [gentle recommendation][1].  Auto-generation will help with scale as we grow the number of types which need these methods, but I'm fine with copy/paste manual generation even then ;).  @abhinavdahiya?

[1]: https://github.com/openshift/installer/issues/256#issuecomment-422492159

Comment (wking) [reply pkg/ipnet/zz_generated.deepcopy.go]: > Can we just use a simplified version of this implementation...

In fact, it looks like `IPNet` will need a manual version of at least `DeepCopyInto` to [avoid][1]:

```
# github.com/openshift/installer/pkg/ipnet
pkg/ipnet/zz_generated.deepcopy.go:26:10: in.IPNet.DeepCopyInto undefined (type net.IPNet has no field or method DeepCopyInto)
```

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/276/pull-ci-openshift-installer-go-vet/294/build-log.txt

Comment (abhinavdahiya) [reply pkg/ipnet/zz_generated.deepcopy.go]: can we do either completely manual or completely generated? Why is DeepCopyInto not generated automatically?

https://github.com/openshift/machine-config-operator/blob/master/Gopkg.toml#L70-L72
generates DeppCopyInto?? 
https://github.com/openshift/machine-config-operator/blob/master/pkg/apis/machineconfiguration.openshift.io/v1/zz_generated.deepcopy.go#L14-L20

**Comment by abhinavdahiya** [comment]
https://github.com/openshift/installer/pull/276#discussion_r218634289

File: pkg/ipnet/zz_generated.deepcopy.go (hunk 1 above)
Comment (wking) [reply pkg/ipnet/zz_generated.deepcopy.go]: > can we do either completely manual or completely generated? Why is `DeepCopyInto` not generated automatically?

Completely manual is an option, although [there are trade-offs][1].  Completly automatic is not, see c3832d290b9dca073c373047dc217afc3925c15a.

[1]: https://github.com/openshift/installer/pull/276#discussion_r218605645

**Comment by dgoodwin** [comment]
For reference you can see how upstream is representing CIDR blocks here: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/cluster_types.go#L63

**Comment by wking** [comment]
> For reference you can see how upstream is representing CIDR blocks here: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/cluster_types.go#L63

I prefer maintaining copy methods (however we decide to do that) to using unstructured strings for subnets.  But I won't fight too hard if everone else wants to go with unstructured strings ;).

**Comment by crawford** [comment]
@wking +1. Let's drop the k8s code generation and JS-esque typing.

**Comment by wking** [comment]
I've pushed c3832d2 -> 270f80e, which drops the code generation and adds two additional `IPNet` cleanups in separate commits.

**Comment by wking** [comment]
Turns out a pointer `MarshalJSON` (692693403e) [breaks our pretty CIDR serialization][1].  I've dropped it with 270f80e -> 06739c6.

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/276/pull-ci-openshift-installer-unit/326/build-log.txt

**Pull request merged**